dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPTX.java

```diff
@@ -0,0 +1,129 @@
+        load.addFilters(scopeFilters);
+        store.addFilters(scopeFilters);
+        Set<String> proxyFilters = getProxyFilters(e);
+        optionalMbBefore.addFilters(proxyFilters);
+        optionalMbAfter.addFilters(proxyFilters);
+        load.addFilters(proxyFilters);
+        store.addFilters(proxyFilters);
+        addSemFilters(e, load);
+        addSemFilters(e, store);
+
+        return eventSequence(
+                optionalMbBefore,
+                load,
+                store,
+                newLocal(resultRegister, dummy),
+                optionalMbAfter
+        );
+    }
+
+    @Override
+    public List<Event> visitRMWOp(RMWOp e) {
+        IExpr address = e.getAddress();
+        Register resultRegister = e.getResultRegister();
+
+        Register dummy = e.getThread().newRegister(resultRegister.getPrecision());
+        Load load = newRMWLoad(dummy, address, Tag.Linux.MO_ONCE);
+        load.addFilters(Tag.Linux.NORETURN);
+        RMWStore store = newRMWStore(load, address,
+                new IExprBin(dummy, e.getOp(), (IExpr) e.getMemValue()), Tag.Linux.MO_ONCE);
+
+        Set<String> scopeFilters = getScopeFilters(e);
+        load.addFilters(scopeFilters);
+        store.addFilters(scopeFilters);
+        Set<String> proxyFilters = getProxyFilters(e);
+        load.addFilters(proxyFilters);
+        store.addFilters(proxyFilters);
+        addSemFilters(e, load);
+        addSemFilters(e, store);
+
+        return eventSequence(
+                load,
+                store
+        );
+    }
+
+    private Set<String> getScopeFilters(Event e) {
+        Set<String> filters = e.getFilters();
+        Set<String> scopeFilters = new HashSet<>();
+        for (String filter : filters) {
+            if (filter.contains(Tag.PTX.CTA) || filter.contains(Tag.PTX.GPU)
+                    || filter.contains(Tag.PTX.SYS)) {
+                scopeFilters.add(filter);
+            }
+        }
+        return scopeFilters;
+    }
+
+    private void addSemFilters(Event e, Load load) {
+        if (e.is(Tag.PTX.ACQ_REL)) {
+            load.addFilters(Tag.PTX.ACQ);
+        } else if (e.is(Tag.PTX.RLX)) {
+            load.addFilters(Tag.PTX.RLX);
+        }
+    }
+
+    private void addSemFilters(Event e, Store store) {
+        if (e.is(Tag.PTX.ACQ_REL)) {
+            store.addFilters(Tag.PTX.REL);
+        } else if (e.is(Tag.PTX.RLX)) {
+            store.addFilters(Tag.PTX.RLX);
+        }
+    }
+
+    private Set<String> getProxyFilters(Event e) {
+        Set<String> filters = e.getFilters();
+        Set<String> scopeFilters = new HashSet<>();
+        for (String filter : filters) {
+            if (filter.equals(Tag.PTX.GEN) || filter.equals(Tag.PTX.TEX) || filter.equals(Tag.PTX.SUR)
+                    || filter.equals(Tag.PTX.CON) || filter.contains(Tag.PTX.ALIAS)) {
+                scopeFilters.add(filter);
+            }
+        }
+        return scopeFilters;
+    }
+
+}
```

Comment:
We can probably get rid of all these methods

Reply:
Accepted